daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala

```diff
@@ -384,6 +384,27 @@ sealed abstract class HasTxNodes[Nid, +Cid] {
       case (acc, _) => acc
     } -- localContracts.keySet
 
+  /** Return all the contract keys references by this transaction.
+    * This includes the keys created, exercised, fetched, or lookup, even those
+    * that refer to transient contracts or that appear under a roolback node.
+    */
+  final def contractKeys(implicit
+      ev: HasTxNodes[Nid, Cid] <:< HasTxNodes[_, Value.ContractId]
+  ): Set[GlobalKey] = {
+    ev(this).fold(Set.empty[GlobalKey]) {
```

Comment:
I'd find this a little prettier with some kind of `foldMap`. Is it worth implementing that?

Reply:
Could be sensible but I’d prefer keeping it for a separate PR.